BUG: Fix dataframe.update not updating dtype #55509 v2

aureliobarbosa opened the pull request:
- [x] closes #55509. 
- [x] 3 Tests added and passed.
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v3.0.rst`. (?)

A new PR was opened to replace the previous one (#55634) because a more idiomatic solution was found. The previous PR will be converted to draft. 

@rhshadrach @mroeschke

aureliobarbosa committed:
TST: add a test for preserving dtype while calling frame.update (#55509)

aureliobarbosa committed:
TST: Add a test for frame.update raising on duplicate argument indexes (#55509)

aureliobarbosa committed:
TST:  frame.update accepts duplicate frame index with unique argument #55509

aureliobarbosa committed:
BUG: fix dataframe.update not preserving dtypes (#55509)

datapythonista commented:
Thanks for the contribution @aureliobarbosa. Can you add a note to the release notes please? You can check other PRs if you are not familiar with it.

datapythonista commented:
> Sure. For which version? 3.0? @datapythonista

Yes please

aureliobarbosa committed:
DOC: Add line indicating bug fix #55509

rhshadrach commented:
Benchmarks from https://github.com/pandas-dev/pandas/pull/55634#discussion_r1375264127

```
size = 100_000
df1 = pd.DataFrame({'a': np.random.randint(0, 100, size)}, index=range(1, size+1))
df2 = pd.DataFrame({'a': np.random.randint(0, 100, 3)}, index=[10, 12, 11])
%timeit df1.update(df2)
# 1.11 ms ± 10.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  <-- PR
# 2.32 ms ± 36.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)    <-- main

size = 100_000
index = list(range(1, size+1))
index[1], index[0] = index[0], index[1]
df1 = pd.DataFrame({'a': np.random.randint(0, 100, size)}, index=index)
df2 = pd.DataFrame({'a': np.random.randint(0, 100, 3)}, index=[10, 12, 11])
%timeit df1.update(df2)
# 1.05 ms ± 3.11 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  <-- PR
# 1.76 ms ± 3.73 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  <-- main

size = 100_000
df1 = pd.DataFrame({'a': np.random.randint(0, 100, 3)}, index=[10, 11, 12])
df2 = pd.DataFrame({'a': np.random.randint(0, 100, size)})
%timeit df1.update(df2)
# 517 µs ± 1.54 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  <-- PR
# 318 µs ± 4.85 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  <-- main

size = 10
df1 = pd.DataFrame({'a': np.random.randint(0, 100, size)})
df2 = pd.DataFrame({'a': [-1, -2, -3]}, index=[4, 4, 6])
df1.update(df2)
# ValueError: Update not allowed with duplicate indexes on other.  <-- PR
# ValueError: cannot reindex on an axis with duplicate labels      <-- main
```

rhshadrach commented on pandas/core/frame.py:
```diff
@@ -8764,11 +8764,22 @@ def update(
 
-        other = other.reindex(self.index)
+        if other.index.has_duplicates:
+            raise ValueError("Update not allowed with duplicate indexes on other.")
```
Can you add a note that this is not supported to the docstring. There isn't currently a notes section, something like this:

https://github.com/pandas-dev/pandas/blob/737d390381ffd8006f15d738c934acb659b5c444/pandas/core/frame.py#L1450-L1454

aureliobarbosa replied:
@rhshadrach Done. 

`DataFrame.reindex` raises on the same condition and does not mention it on the docs. What about adding a similar note there?

rhshadrach replied:
Slight preference on keeping this PR just modifying update, but no strong objection to modifying the reindex docs.

aureliobarbosa replied:
Ok. It can be done in another PR.

rhshadrach reviewed (changes requested):
lgtm - one request

aureliobarbosa committed:
DOC: add note on duplicate indices on parameter other (#55509)

rhshadrach commented:
/preview

rhshadrach reviewed (approved):
lgtm

datapythonista commented on pandas/core/frame.py:
```diff
@@ -8692,6 +8692,10 @@ def update(
         DataFrame.merge : For column(s)-on-column(s) operations.
 
+        Notes
+        --------
```
I think having the `-------` longer than the title used to make the CI crash. Not sure why the CI is happy now, but maybe worth making `Notes` and it the same length in case it fails again in the future.

aureliobarbosa replied:
Now it has the same size.

datapythonista commented on pandas/core/frame.py:
```diff
@@ -8764,11 +8768,22 @@ def update(
+        rows = other.index.intersection(self.index)
+        if rows.empty:
+            raise ValueError(
+                "Can't update dataframe when other has no index in common with "
```
No big deal, but in the previous error message you say "update not allowed" and here "can't update dataframe" to mean the same thing. Not sure how we usually write error messages, but I think it'd be better to be consistent.

datapythonista commented on pandas/tests/frame/methods/test_update.py:
```diff
@@ -184,3 +184,47 @@ def test_update_dt_column_with_NaT_create_column(self):
+    )
+    def test_update_preserve_dtype(self, value_df, value_other, dtype):
+        # GH#55509
+        df = DataFrame({"a": [value_df] * 2}, index=[1, 2])
```
You don't seem to use the `dtype` parameter in any fixture. Did you forget to specify it in the constructors, or am I missing something?

aureliobarbosa replied:
Thanks for the review and suggestions @datapythonista . 

I will address all points as soon as possible (hope to finish by tomorrow).

Regarding the tests, note that tm.assert_frame_equal does assert that everything in the frame is equal. By the way, I checked this behavior for values and dtypes (didn't look indexes and column's labels).

datapythonista replied:
My question is, why do you have a parameter for the `dtype` in the pytest parametrization, if it's not being used? I had the impression that you are planning to set the dtype explicitly when creating the dataframe, but maybe you forgot. If it's not needed, maybe better to remove from the pytest parameters?

aureliobarbosa replied:
@datapythonista 
You are absolutely right!

My tests were originally base on the [tests](https://github.com/pandas-dev/pandas/issues/55509#issuecomment-1767324695) proposed by @MichaelTiemannOSC. I tried to turn them simpler and similar to other tests in the same test set, but since they were failing on main and passing this PR I forgot about them. 

This is addressed in the last two commits, which make two of the tests that I propose strictier. Nothing changes but those tests are better now! Thanks for the guidance on this point.

aureliobarbosa replied:
Complementing, I think it is better to let the `dtypes` to be explicitly stated on those cases. The reason is that in this way it is possible to check `dtype` consistency for native python `int`, numpy, and Extended Arrays, as expected through the 'parametrization'.

datapythonista reviewed (commented):
Looks great, thanks for working on this. I added couple of comments.

Also, I was checking that we don't seem to have an automated benchmark to see if `DataFrame.update` is taking longer after this change. I don't think there should be a big impact, but maybe you can check locally how a sample case performs before and after this change (e.g. `%timeit df.update(other_df)`). Also, you are very welcome to add a benchmark (if you do it, maybe you can open another PR, and we can merge it first, so we can see if there is a regression here).

Thanks!

aureliobarbosa committed:
TST:  assure test_update_preserve_dtype  checks dtype

aureliobarbosa committed:
TST: assure test_update_on_duplicate_frame_unique_argument_index checks dtype

aureliobarbosa committed:
TST: Use np.intc instead of np.int64 on dtype tests for frame.update (#55509)

aureliobarbosa committed:
Merge remote-tracking branch 'upstream/main' into fix_gh55509_v2

aureliobarbosa committed:
Merge remote-tracking branch 'upstream/main' into fix_gh55509_v2

aureliobarbosa committed:
DOC: Fix separator size (#55509). Minor issue.

aureliobarbosa committed:
DOC: fix error messages on frame.update (PR #57637)

aureliobarbosa commented:
@datapythonista I am looking into ASV right now. Which kind of benchmark would be more useful: time, memory or both? Can you point me to a properly designed pandas benchmark so that I can have a reference for an start on this?

Independent of this, manual tests are going to be done locally and posted here for reference, in the following.

datapythonista reviewed (approved):
Do you think it's worth adding a use case to test when the indices have no intersection?

Also, no big deal, but not sure if we can find a better name for the variable `rows`. Maybe `mask` or `index_intersection` is better.

In any case, this looks good to me.

datapythonista commented:
I think all our benchmarks should be reasonable. Understanding and running asv may be a bit tricky, but writing the benchmark should be trivial, just create some sample data and call `DataFrame.update` with it. I think checking time is probably good enough. The main complexity in my opinion is to use enough data so the timing is not super noise, but not have too much so the benchmark is slow.

Also, we use benchmarks more for the algorithms that we implement ourselves (like in C) that to check the performance of everything.  I think in this case it can be useful, but use your own judgement to decide whether it's useful or not.

aureliobarbosa committed:
TST:  add test_update_raises_without_intersection on DataFrame (#55509).

aureliobarbosa committed:
Rename variable.

aureliobarbosa commented:
@datapythonista 

> Do you think it's worth adding a use case to test when the indices have no intersection?

I agree because this is a behavior that was introduced on this PR. Added the test for this case and changed the message raise slightly to emphasize 'no intersection'.

> not sure if we can find a better name for the variable rows. Maybe mask or index_intersection is better.

The first of this PR used `indexes_intersection`, but then I changed to `rows` to simplify the notation. Now that you raised this point, I noted that a simple search on the file `frame.py` results in more than one hundred matches. Considering this, I agree with avoided using `rows`.

aureliobarbosa commented:
I did benchmark again the first three manual code tests proposed previously by @rhshadrach. Results are summarized below in the same order.

| Data | main  | PR  | 
| --- | --- | --- |
| monotonic  |  1.8 ms  | 1.29  ms  |  
| non-monotonic |  1.83 ms | 1.21 ms  |  
| small-frame-large-arg | 384 µs |595 µs | 

The performance gain is not as good as originally expected, but they are still relevant. @datapythonista Note that when the frame is small and the argument is large there is a performance loss, but as pointed out previously by @rhshadrach  this looks like an edge use case.

aureliobarbosa commented:
Hope to be done here. 

@datapythonista Next thing I will do is to try to implement an ASV benchmark for this function.

@rhshadrach I think that the take home message on this PR was that the function was manipulating data it wasn't expecting to touch on, and this case, the change in the dtype was the symptom for this problem.  

If you both know or find any bug that could be related or similar to this one, I would be glad to contribute.

Regards

datapythonista reviewed (approved):
Excellent work @aureliobarbosa.

Do you want to have a last look @rhshadrach?

rhshadrach reviewed (approved):
lgtm

rhshadrach commented:
Thanks @aureliobarbosa!
